rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,1039 @@
+      }
+    );
+    if (gid_seq.end() != it) {
+      return {node_info.node_name, node_info.node_namespace, EndpointCreator::ROS_NODE};
+    }
+  }
+  return {"", "", EndpointCreator::UNDISCOVERED_ROS_NODE};
+}
+
+using DemangleFunctionT = GraphCache::DemangleFunctionT;
+
+static
+rmw_ret_t
+__get_entities_info_by_topic(
+  const GraphCache::EntityGidToInfo & entities,
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  const std::string & topic_name,
+  DemangleFunctionT demangle_type,
+  bool is_reader,
+  rcutils_allocator_t * allocator,
+  rmw_topic_endpoint_info_array_t * endpoints_info)
+{
+  assert(allocator);
+  assert(endpoints_info);
+
+  if (0u == entities.size()) {
+    return RMW_RET_OK;
+  }
+
+  size_t size = std::count_if(
+    entities.begin(),
+    entities.end(),
+    [&topic_name](const GraphCache::EntityGidToInfo::value_type & item) {
+      return item.second.topic_name == topic_name;
+    }
+  );
+  if (0u == size) {
+    return RMW_RET_OK;
+  }
+
+  rmw_ret_t ret = rmw_topic_endpoint_info_array_init_with_size(
+    endpoints_info,
+    size,
+    allocator);
+  // TODO(ivanpauno): `count` should be initialized by
+  //  `rmw_topic_endpoint_info_array_init_with_size`, and should be probably called `size`.
+  endpoints_info->count = size;
+  if (RMW_RET_OK != ret) {
+    return ret;
+  }
+  std::unique_ptr<
+    rmw_topic_endpoint_info_array_t,
+    std::function<void(rmw_topic_endpoint_info_array_t *)>>
+  endpoints_info_delete_on_error(
+    endpoints_info,
+    [allocator](rmw_topic_endpoint_info_array_t * p) {
+      rmw_ret_t ret = rmw_topic_endpoint_info_array_fini(
+        p,
+        allocator
+      );
+      if (RMW_RET_OK != ret) {
+        RCUTILS_LOG_ERROR_NAMED(
+          log_tag,
+          "Failed to destroy endpoints_info when function failed.");
+      }
+    }
+  );
+
+  size_t i = 0;
+  for (const auto & entity_pair : entities) {
+    if (entity_pair.second.topic_name != topic_name) {
+      continue;
+    }
+
+    rmw_topic_endpoint_info_t & endpoint_info = endpoints_info->info_array[i];
+    endpoint_info = rmw_get_zero_initialized_topic_endpoint_info();
+
+    auto result = __find_name_and_namespace_from_entity_gid(
+      participant_map,
+      entity_pair.second.participant_gid,
+      entity_pair.first,
+      is_reader);
+
+    std::string node_name;
+    std::string node_namespace;
+    switch (std::get<2>(result)) {
+      case EndpointCreator::ROS_NODE:
+        node_name = std::move(std::get<0>(result));
+        node_namespace = std::move(std::get<1>(result));
+        break;
+      case EndpointCreator::UNDISCOVERED_ROS_NODE:
+        node_name = "_NODE_NAME_UNKNOWN_";
+        node_namespace = "_NODE_NAMESPACE_UNKNOWN_";
+        break;
+      case EndpointCreator::BARE_DDS_PARTICIPANT:
+        node_name = "_CREATED_BY_BARE_DDS_APP_";
+        node_namespace = "_CREATED_BY_BARE_DDS_APP_";
```

Comment:
@ivanpauno why not using these string literals when first constructing the tuple? Then you can optimize this `switch` away.

Reply:
I could optimize `EndpointCreator::BARE_DDS_PARTICIPANT` case, but not also `EndpointCreator::UNDISCOVERED_ROS_NODE`. I find clearer if all the cases are together.